bootique-shiro-web-oidconnect/src/main/java/io/bootique/shiro/web/oidconnect/JwtOpenIdCallbackHandler.java

```diff
@@ -0,0 +1,185 @@
+package io.bootique.shiro.web.oidconnect;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import jakarta.servlet.ServletRequest;
+import jakarta.ws.rs.ForbiddenException;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.client.ClientBuilder;
+import jakarta.ws.rs.client.Entity;
+import jakarta.ws.rs.client.WebTarget;
+import jakarta.ws.rs.core.*;
+import org.eclipse.jetty.http.HttpStatus;
+import org.glassfish.jersey.client.JerseyClient;
+import org.glassfish.jersey.client.JerseyClientBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.external.biz.base64Coder.Base64Coder;
+
+import java.net.URI;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+@Path("/bq-shiro-oauth-callback")
+public class JwtOpenIdCallbackHandler implements OidConnect {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JwtOpenIdCallbackHandler.class);
+
+    private final ObjectMapper mapper;
+
+    private final JerseyClient webClient;
+    private final WebTarget tokenTarget;
+    private final String tokenCookie;
+    private final String clientId;
+    private final String clientSecretKey;
+    private final String oidpUrl;
+    private final String callbackUri;
+    private final String scope;
+
+    public JwtOpenIdCallbackHandler(ObjectMapper objectMapper,
+                                    String tokenCookie,
+                                    String tokenUrl,
+                                    String clientId,
+                                    String clientSecretKey,
+                                    String scope,
+                                    String oidpUrl,
+                                    String callbackUri) {
+        this.mapper = objectMapper;
+        this.tokenCookie = tokenCookie;
+        this.clientId = clientId;
+        this.clientSecretKey = URLEncoder.encode(clientSecretKey, StandardCharsets.UTF_8);
+        this.scope = scope;
+        this.webClient = JerseyClientBuilder.createClient();
+        this.tokenTarget = webClient.target(tokenUrl);
+        this.oidpUrl = oidpUrl;
+        this.callbackUri = callbackUri;
+    }
+
+    private Form form(String code) {
+        Form form = new Form()
+                .param(GRANT_TYPE_PARAMETER_NAME, GRANT_TYPE_AUTH_CODE_VALUE)
+                .param(CLIENT_ID_PARAMETER_NAME, clientId)
+                .param(CLIENT_SECRET_KEY_PARAMETER_NAME, clientSecretKey)
+                .param(CODE_PARAMETER_NAME, code);
+        if (scope != null && !scope.isEmpty()) {
+            form = form.param(SCOPE_PARMETER_NAME, scope);
+        }
+        return form;
+    }
+
+    @GET
+    public Response callback(@Context UriInfo uriInfo) {
```

Comment:
We know the parameter names, so we can model them in the method signature using `@QueryParam`